File: PraticasIAADSP2/src/main.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <unordered_map>
#include <vector>

struct DataRow {
  std::string country;
  std::string region;
  int rank_felicidade;
  double score_felicidade;
  double stand_error;
  double PIB;
  double family;
  double expect_vida;
  double freedom;
  double currupcao;
  double generosity;
  double dystopia;
  int country_code;
};

std::vector<std::string> split(const std::string &s, char delimiter) {
  std::vector<std::string> tokens;
  std::string token;
  std::istringstream tokenStream(s);
  while (std::getline(tokenStream, token, delimiter)) {
    tokens.push_back(token);
  }
  return tokens;
}

int main() {
  std::ifstream file("database/mundo_feliz.csv");
  std::string line;
  std::vector<DataRow> dataset;

  if (!file.is_open()) {
    std::cerr << "Erro ao abrir o arquivo!" << std::endl;
    return 1;
  }

  // Lê o cabeçalho
  std::getline(file, line);

  std::unordered_map<std::string, int> country_codes;
  int next_code = 0;

  while (std::getline(file, line)) {
    auto cols = split(line, ',');
    if (cols.size() != 12)
      continue;

    DataRow row;
    row.country = cols[0];
    row.region = cols[1];
    row.rank_felicidade = std::stoi(cols[2]);
    row.score_felicidade = std::stod(cols[3]);
    row.stand_error = std::stod(cols[4]);
    row.PIB = std::stod(cols[5]);
    row.family = std::stod(cols[6]);
    row.expect_vida = std::stod(cols[7]);
    row.freedom = std::stod(cols[8]);
    row.currupcao = std::stod(cols[9]);
    row.generosity = std::stod(cols[10]);
    row.dystopia = std::stod(cols[11]);

    // Converter país em categoria
    if (country_codes.find(row.country) == country_codes.end()) {
      country_codes[row.country] = next_code++;
    }
    row.country_code = country_codes[row.country];

    dataset.push_back(row);
  }

  // Exibe info geral
  std::cout << "Total de linhas: " << dataset.size() << std::endl;
  std::cout << "Total de colunas: 12 + 1 (country_code)" << std::endl;
  std::cout << "Primeiros 5 registros:\n";

  for (size_t i = 0; i < std::min(dataset.size(), size_t(5)); ++i) {
    const auto &row = dataset[i];
    std::cout << row.country << " (" << row.country_code << "), " << row.region
              << ", Rank: " << row.rank_felicidade
              << ", Score: " << row.score_felicidade << "\n";
  }

  return 0;
}

File: PraticasIAADSP2/src/main.py
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os

# Caminho do CSV
file_path = './database/mundo_feliz.csv'

# Verifica se o arquivo existe
if not os.path.exists(file_path):
    print(f"Arquivo não encontrado: {file_path}")
    exit()

# Lê o CSV
df = pd.read_csv(file_path)

# Renomeia colunas
df.columns = ['Country', 'Region', 'rank_felicidade', 'score_felicidade',
              'stand_error', 'PIB', 'Family', 'expect_vida', 'Freedom',
              'currupcao', 'Generosity', 'Dystopia Residual']

# Estilo dos gráficos
sns.set(style="darkgrid")
plt.rcParams['figure.figsize'] = (10, 6)

# Gráfico 1 - Shape do DataFrame (tamanho de linhas e colunas)
plt.figure()
plt.bar(['Linhas', 'Colunas'], df.shape, color='slateblue')
plt.title("Shape do DataFrame (linhas e colunas)")
plt.ylabel("Quantidade")
plt.tight_layout()
plt.show()

# Gráfico 2 - Valores nulos por coluna
plt.figure()
df.isnull().sum().plot(kind='bar', color='darkred')
plt.title("Valores Nulos por Coluna")
plt.ylabel("Quantidade de Nulos")
plt.xlabel("Colunas")
plt.xticks(rotation=45)
plt.tight_layout()
plt.show()

# Gráfico 3 - Histograma do Rank de F elicidade
plt.figure()
plt.hist(df['rank_felicidade'], bins=20, color='orange', edgecolor='black')
plt.title("Distribuição do Rank de Felicidade")
plt.xlabel("Rank")
plt.ylabel("Frequência")
plt.tight_layout()
plt.show()

# Gráfico 4 - Estatísticas Descritivas (exemplo: expectativa de vida)
expect = df['expect_vida'].describe()
plt.figure()
plt.bar(expect.index, expect.values, color='green')
plt.title("Estatísticas da Expectativa de Vida")
plt.ylabel("Valor")
plt.xticks(rotation=45)
plt.tight_layout()
plt.show()
print("Testando a versão do pandas")

plt.tight_layout()
plt.show()

# Gráfico 6 - Score de Felicidade (Histograma)
plt.figure()
plt.hist(df['score_felicidade'], bins=15, color='deepskyblue', edgecolor='black')
plt.title("Distribuição do Score de Felicidade")
plt.xlabel("Score")
plt.ylabel("Frequência")
plt.tight_layout()
plt.show()

# Gráfico 7 - Correlação entre PIB e Expectativa de Vida (Scatter)
plt.figure()
plt.scatter(df['PIB'], df['expect_vida'], alpha=0.7, color='teal')
plt.title("PIB vs Expectativa de Vida")
plt.xlabel("PIB")
plt.ylabel("Expectativa de Vida")
plt.grid(True)
plt.tight_layout()
plt.show()

# Gráfico 8 - Média, Mediana e Máximo da Expectativa de Vida
valores = {
    'Média': df['expect_vida'].mean(),
    'Mediana': df['expect_vida'].median(),
    'Máximo': df['expect_vida'].max()
}

plt.figure()
plt.bar(valores.keys(), valores.values(), color=['cyan', 'magenta', 'yellow'])
plt.title("Expectativa de Vida - Média, Mediana e Máximo")
plt.ylabel("Expectativa de Vida")
plt.tight_layout()
plt.show()
